core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveFalseFiltersAfterDomainTranslator.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.metadata.Metadata;
+import io.trino.sql.PlannerContext;
+import io.trino.sql.planner.DomainTranslator;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import io.trino.sql.tree.Expression;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.trino.sql.ExpressionUtils.combineConjuncts;
+import static io.trino.sql.ExpressionUtils.extractConjuncts;
+import static io.trino.sql.planner.DeterminismEvaluator.isDeterministic;
+import static io.trino.sql.planner.plan.Patterns.filter;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Uses DomainTranslator#getExtractionResult to infer that the expression is "false" in some cases (TupleDomain.none()).
+ * <p>
+ * Transforms
+ * <pre>
+ *  - Filter (FALSE_EXPRESSION)
+ * </pre>
+ * into
+ * <pre>
+ *  - Values (0)
+ * </pre>
+ */
+public class RemoveFalseFiltersAfterDomainTranslator
```

Comment:
I think I have commented about that previously, but cannot find this now. Sorry if I repeat myself or if you already have responded to that.

I think the new rule should extend `ExpressionRewriteRuleSet` and thus apply to any boolean expression that can be proven to be false. This would apply to filternode, but also to projections (IF, CASE or just projecting the result of a boolean expression). Then, combined with RemoveTrivialFilters, would result in more trivial filters being eliminated.

Reply:
Sounds good, done

Reply:
I'm getting the following failure:
```
Caused by: java.lang.IllegalArgumentException: Function is not resolved: count
	at io.trino.metadata.MetadataManager.lambda$decodeFunction$49(MetadataManager.java:2093)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at io.trino.metadata.MetadataManager.decodeFunction(MetadataManager.java:2093)
	at io.trino.sql.planner.DeterminismEvaluator.lambda$isDeterministic$0(DeterminismEvaluator.java:37)
	at io.trino.sql.planner.DeterminismEvaluator$Visitor.visitFunctionCall(DeterminismEvaluator.java:63)
	at io.trino.sql.planner.DeterminismEvaluator$Visitor.visitFunctionCall(DeterminismEvaluator.java:50)
	at io.trino.sql.tree.FunctionCall.accept(FunctionCall.java:121)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.planner.DeterminismEvaluator.isDeterministic(DeterminismEvaluator.java:46)
	at io.trino.sql.planner.DeterminismEvaluator.isDeterministic(DeterminismEvaluator.java:37)
	at io.trino.sql.planner.iterative.rule.RemoveFalseFiltersAfterDomainTranslator.rewrite(RemoveFalseFiltersAfterDomainTranslator.java:55)
```
on `SELECT count(*) FROM tpch.tiny.nation` (see https://github.com/trinodb/trino/actions/runs/4588988784/jobs/8103600850?pr=16206).
I'm adding a try-catch but I'm not sure that's the right thing to do. Another option is to create a new visitor \ change `DeterminismEvaluator`. WDYT?

Reply:
Another issue - we are now translating any `CAST(null AS <type>)` to `FALSE` although when not within a filter, the expression sometimes should stay null. For example, run `io.trino.plugin.postgresql.TestPostgreSqlTableStatistics#testNullsFraction`

Reply:
> correct.

So I shall revert this back to a regular rule, right?

> is postgresql test the only one catching that?
> please add something to AbstractTestEngineOnlyQueries as well

`AbstractTestEngineOnlyQueries` already catch this at many tests, for example `io.trino.testing.AbstractTestEngineOnlyQueries#testNullOnLhsOfInPredicateAllowed` fails at 
`assertQuery("SELECT NULL IN (1, 2, 3)", "SELECT NULL");`
on
```
java.lang.AssertionError: For query 20230413_074905_00696_vgf9n: 
 SELECT NULL IN (1, 2, 3)
not equal
Actual rows (up to 100 of 1 extra rows shown, 1 rows in total):
    [false]
Expected rows (up to 100 of 1 missing rows shown, 1 rows in total):
    [null]
```

Reply:
> do you see cases other than filtering where converting to `FALSE` is the right thing to do?
> 
> >

`IF(an-expression-that-cannot-be-true, foo, bar)` -> `IF(false, foo, bar)` -> `bar`

Reply:
> IF(an-expression-that-cannot-be-true, foo, bar) -> IF(false, foo, bar) -> bar

This doesn't seem to be a valid query
```
trino> select if(CAST(NULL AS BIGINT), 'a', 'b');
Query 20230330_045558_00017_npitn failed: line 1:11: IF condition must evaluate to a boolean (actual: bigint)
select if(CAST(NULL AS BIGINT), 'a', 'b')

```

> > So shall I revert this back to a regular rule?
> 
> cc @martint

I came to a realization that `joinExpressionRewrite` can also be executed. In general, we can apply this logic only on expressions that are used for filtering, otherwise `TupleDomain.none()` might represent a `NULL` value. So there's no reason to revert. See updated implementation.

Reply:
> after switching to `ExpressionRewriteRuleSet` "RemoveFalseFiltersAfterDomainTranslator" is no longer a good rule name.
> 
> "SimplifyFalseConditions"

Done.

Reply:
> This doesn't seem to be a valid query
> 
> ```
> trino> select if(CAST(NULL AS BIGINT), 'a', 'b');
> ```

what if you cast it to `boolean`?

Reply:
> I came to a realization that joinExpressionRewrite can also be executed. In general, we can apply this logic only on expressions that are used for filtering, otherwise TupleDomain.none() might represent a NULL value. So there's no reason to revert. I've removed all rewrites except filterExpressionRewrite and joinExpressionRewrite .

that's a big more involved

for example

in projection , `CAST(NULL AS boolean)` should be retained
 in filter, `CAST(NULL AS boolean)` can be replaced with `false`

but

in projection, `IF(CAST(NULL AS boolean), 'a', 'b')` can be replaced with `IF(false, 'a', 'b')` and then to `'b'`
in filter, `CAST(NULL AS boolean) IS NULL` cannot be replaced with `FALSE IS NULL`

Reply:
> > This doesn't seem to be a valid query
> > ```
> > trino> select if(CAST(NULL AS BIGINT), 'a', 'b');
> > ```
> 
> what if you cast it to `boolean`?

Right, with `boolean` it's legit. I've run the following query - `SELECT * FROM nation WHERE if(CAST(NULL AS BOOLEAN), TRUE, FALSE)` and saw that `io.trino.sql.planner.iterative.rule.SimplifyExpressions#rewrite` (which calls `ExpressionInterpreter`) rewrites `(CASE WHEN CAST(null AS BOOLEAN) THEN true ELSE false END)` into `false` so it's already covered.


> in projection, IF(CAST(NULL AS boolean), 'a', 'b') can be replaced with IF(false, 'a', 'b') and then to 'b'

The same thing happens in project, I've run the query `SELECT if(CAST(NULL AS BOOLEAN), TRUE, FALSE) FROM nation` and `io.trino.sql.planner.iterative.rule.SimplifyExpressions#rewrite` rewrited the expression into 'false'.


> in filter, CAST(NULL AS boolean) IS NULL cannot be replaced with FALSE IS NULL

AFAIU such expressions should be handled by the domain translator (via `visitIsNullPredicate` in this example) otherwise https://github.com/trinodb/trino/commit/55ab1ec7ecb3b5c7a5ab5b829877cab42603a14c#diff-9d28eac8af7badcb977041a5afc38363315d66db73a8556a1422e21e9233b63fR481-R483 is erroneous.